# S12 T01: Aprenentatge Supervisat - Regressions

## Nivell 1

### *Exercici 1 + Exercici 2*

Crea almenys tres models de regressió diferents per intentar predir el millor possible l’endarreriment dels vols (ArrDelay) de DelayedFlights.csv.

Compara’ls en base al MSE i al R2 .

In [1]:
import numpy as np
import pandas as pd
from pandas import read_csv
from sklearn.model_selection import train_test_split
%matplotlib inline
import matplotlib.pyplot as plt

In [2]:
retards_df = pd.read_csv('DelayedFlights.csv')

In [3]:
retards_df

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0                 0  2008      1           3          4   2003.0        1955   
1                 1  2008      1           3          4    754.0         735   
2                 2  2008      1           3          4    628.0         620   
3                 4  2008      1           3          4   1829.0        1755   
4                 5  2008      1           3          4   1940.0        1915   
...             ...   ...    ...         ...        ...      ...         ...   
1936753     7009710  2008     12          13          6   1250.0        1220   
1936754     7009717  2008     12          13          6    657.0         600   
1936755     7009718  2008     12          13          6   1007.0         847   
1936756     7009726  2008     12          13          6   1251.0        1240   
1936757     7009727  2008     12          13          6   1110.0        1103   

         ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
0         2211.0        2225            WN  ...     4.0     8.0          0   
1         1002.0        1000            WN  ...     5.0    10.0          0   
2          804.0         750            WN  ...     3.0    17.0          0   
3         1959.0        1925            WN  ...     3.0    10.0          0   
4         2121.0        2110            WN  ...     4.0    10.0          0   
...          ...         ...           ...  ...     ...     ...        ...   
1936753   1617.0        1552            DL  ...     9.0    18.0          0   
1936754    904.0         749            DL  ...    15.0    34.0          0   
1936755   1149.0        1010            DL  ...     8.0    32.0          0   
1936756   1446.0        1437            DL  ...    13.0    13.0          0   
1936757   1413.0        1418            DL  ...     8.0    11.0          0   

         CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
0                       N         0           NaN           NaN      NaN   
1                       N         0           NaN           NaN      NaN   
2                       N         0           NaN           NaN      NaN   
3                       N         0           2.0           0.0      0.0   
4                       N         0           NaN           NaN      NaN   
...                   ...       ...           ...           ...      ...   
1936753                 N         0           3.0           0.0      0.0   
1936754                 N         0           0.0          57.0     18.0   
1936755                 N         0           1.0           0.0     19.0   
1936756                 N         0           NaN           NaN      NaN   
1936757                 N         0           NaN           NaN      NaN   

        SecurityDelay  LateAircraftDelay  
0                 NaN                NaN  
1                 NaN                NaN  
2                 NaN                NaN  
3                 0.0               32.0  
4                 NaN                NaN  
...               ...                ...  
1936753           0.0               22.0  
1936754           0.0                0.0  
1936755           0.0               79.0  
1936756           NaN                NaN  
1936757           NaN                NaN  

[1936758 rows x 30 columns]

In [4]:
retards_df.isnull().sum()

Unnamed: 0                0
Year                      0
Month                     0
DayofMonth                0
DayOfWeek                 0
DepTime                   0
CRSDepTime                0
ArrTime                7110
CRSArrTime                0
UniqueCarrier             0
FlightNum                 0
TailNum                   5
ActualElapsedTime      8387
CRSElapsedTime          198
AirTime                8387
ArrDelay               8387
DepDelay                  0
Origin                    0
Dest                      0
Distance                  0
TaxiIn                 7110
TaxiOut                 455
Cancelled                 0
CancellationCode          0
Diverted                  0
CarrierDelay         689270
WeatherDelay         689270
NASDelay             689270
SecurityDelay        689270
LateAircraftDelay    689270
dtype: int64

+ *Comprovades les variables que contenen null values, per tal de conservar el màxim d'observacions/dades que ofereix el 
   dataset; decideixo substituïr-los per la mean de cada una de les variables*

In [5]:
air_time_mean = retards_df['ArrTime'].mean()
retards_df['ArrTime'].fillna(air_time_mean, inplace=True)

In [6]:
retards_df['TailNum'].fillna(method ='pad', inplace = True)

In [7]:
aelaps_time_mean = retards_df['ActualElapsedTime'].mean()
retards_df['ActualElapsedTime'].fillna(air_time_mean, inplace=True)

In [8]:
crseelaps_time_mean = retards_df['CRSElapsedTime'].mean()
retards_df['CRSElapsedTime'].fillna(crseelaps_time_mean, inplace=True)

In [9]:
crseelaps_time_mean = retards_df['AirTime'].mean()
retards_df['AirTime'].fillna(crseelaps_time_mean, inplace=True)

In [10]:
arrdelay_time_mean = retards_df['ArrDelay'].mean()
retards_df['ArrDelay'].fillna(arrdelay_time_mean, inplace=True)

In [11]:
taxiin_time_mean = retards_df['TaxiIn'].mean()
retards_df['TaxiIn'].fillna(taxiin_time_mean, inplace=True)

In [12]:
taxiout_time_mean = retards_df['TaxiOut'].mean()
retards_df['TaxiOut'].fillna(taxiout_time_mean, inplace=True)

In [13]:
carrierdelay_time_mean = retards_df['CarrierDelay'].mean()
retards_df['CarrierDelay'].fillna(carrierdelay_time_mean, inplace=True)

In [14]:
weatherdelay_time_mean = retards_df['WeatherDelay'].mean()
retards_df['WeatherDelay'].fillna(weatherdelay_time_mean, inplace=True)

In [15]:
nasdelay_time_mean = retards_df['NASDelay'].mean()
retards_df['NASDelay'].fillna(nasdelay_time_mean, inplace=True)

In [16]:
securitydelay_time_mean = retards_df['SecurityDelay'].mean()
retards_df['SecurityDelay'].fillna(securitydelay_time_mean, inplace=True)

In [17]:
aircraftdelay_time_mean = retards_df['LateAircraftDelay'].mean()
retards_df['LateAircraftDelay'].fillna(aircraftdelay_time_mean, inplace=True)

In [18]:
retards_df

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0                 0  2008      1           3          4   2003.0        1955   
1                 1  2008      1           3          4    754.0         735   
2                 2  2008      1           3          4    628.0         620   
3                 4  2008      1           3          4   1829.0        1755   
4                 5  2008      1           3          4   1940.0        1915   
...             ...   ...    ...         ...        ...      ...         ...   
1936753     7009710  2008     12          13          6   1250.0        1220   
1936754     7009717  2008     12          13          6    657.0         600   
1936755     7009718  2008     12          13          6   1007.0         847   
1936756     7009726  2008     12          13          6   1251.0        1240   
1936757     7009727  2008     12          13          6   1110.0        1103   

         ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
0         2211.0        2225            WN  ...     4.0     8.0          0   
1         1002.0        1000            WN  ...     5.0    10.0          0   
2          804.0         750            WN  ...     3.0    17.0          0   
3         1959.0        1925            WN  ...     3.0    10.0          0   
4         2121.0        2110            WN  ...     4.0    10.0          0   
...          ...         ...           ...  ...     ...     ...        ...   
1936753   1617.0        1552            DL  ...     9.0    18.0          0   
1936754    904.0         749            DL  ...    15.0    34.0          0   
1936755   1149.0        1010            DL  ...     8.0    32.0          0   
1936756   1446.0        1437            DL  ...    13.0    13.0          0   
1936757   1413.0        1418            DL  ...     8.0    11.0          0   

         CancellationCode  Diverted  CarrierDelay  WeatherDelay   NASDelay  \
0                       N         0     19.179399      3.703571  15.021635   
1                       N         0     19.179399      3.703571  15.021635   
2                       N         0     19.179399      3.703571  15.021635   
3                       N         0      2.000000      0.000000   0.000000   
4                       N         0     19.179399      3.703571  15.021635   
...                   ...       ...           ...           ...        ...   
1936753                 N         0      3.000000      0.000000   0.000000   
1936754                 N         0      0.000000     57.000000  18.000000   
1936755                 N         0      1.000000      0.000000  19.000000   
1936756                 N         0     19.179399      3.703571  15.021635   
1936757                 N         0     19.179399      3.703571  15.021635   

        SecurityDelay  LateAircraftDelay  
0            0.090137          25.296466  
1            0.090137          25.296466  
2            0.090137          25.296466  
3            0.000000          32.000000  
4            0.090137          25.296466  
...               ...                ...  
1936753      0.000000          22.000000  
1936754      0.000000           0.000000  
1936755      0.000000          79.000000  
1936756      0.090137          25.296466  
1936757      0.090137          25.296466  

[1936758 rows x 30 columns]

+ *I comprovo que ja no hi ha cap null value*

In [19]:
retards_df.isnull().sum()

Unnamed: 0           0
Year                 0
Month                0
DayofMonth           0
DayOfWeek            0
DepTime              0
CRSDepTime           0
ArrTime              0
CRSArrTime           0
UniqueCarrier        0
FlightNum            0
TailNum              0
ActualElapsedTime    0
CRSElapsedTime       0
AirTime              0
ArrDelay             0
DepDelay             0
Origin               0
Dest                 0
Distance             0
TaxiIn               0
TaxiOut              0
Cancelled            0
CancellationCode     0
Diverted             0
CarrierDelay         0
WeatherDelay         0
NASDelay             0
SecurityDelay        0
LateAircraftDelay    0
dtype: int64

In [20]:
retards_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1936758 entries, 0 to 1936757
Data columns (total 30 columns):
 #   Column             Dtype  
---  ------             -----  
 0   Unnamed: 0         int64  
 1   Year               int64  
 2   Month              int64  
 3   DayofMonth         int64  
 4   DayOfWeek          int64  
 5   DepTime            float64
 6   CRSDepTime         int64  
 7   ArrTime            float64
 8   CRSArrTime         int64  
 9   UniqueCarrier      object 
 10  FlightNum          int64  
 11  TailNum            object 
 12  ActualElapsedTime  float64
 13  CRSElapsedTime     float64
 14  AirTime            float64
 15  ArrDelay           float64
 16  DepDelay           float64
 17  Origin             object 
 18  Dest               object 
 19  Distance           int64  
 20  TaxiIn             float64
 21  TaxiOut            float64
 22  Cancelled          int64  
 23  CancellationCode   object 
 24  Diverted           int64  
 25  CarrierDelay      

+ *Em desfaig de les columnes amb valors no numèrics, de "Unnamed: 0" que sembla un índex afegit al del DataFrame, de
   "Year" que és el mateix per totes les observacions (2008), de les variables "Time" que recollen hores com a íntegers o 
   floats; i de "Cancelled" i "Diverted" que són variables categòriques convertides i tractades de forma binaria(0, 1)*

In [21]:
retards_df.drop(retards_df.columns[[0, 1, 5, 6, 7, 8,  9, 11, 17, 18, 22, 23, 24]], axis = 1, inplace = True)

In [22]:
retards_df

Month  DayofMonth  DayOfWeek  FlightNum  ActualElapsedTime  \
0            1           3          4        335              128.0   
1            1           3          4       3231              128.0   
2            1           3          4        448               96.0   
3            1           3          4       3920               90.0   
4            1           3          4        378              101.0   
...        ...         ...        ...        ...                ...   
1936753     12          13          6       1621              147.0   
1936754     12          13          6       1631              127.0   
1936755     12          13          6       1631              162.0   
1936756     12          13          6       1639              115.0   
1936757     12          13          6       1641              123.0   

         CRSElapsedTime  AirTime  ArrDelay  DepDelay  Distance  TaxiIn  \
0                 150.0    116.0     -14.0       8.0       810     4.0   
1                 145.0    113.0       2.0      19.0       810     5.0   
2                  90.0     76.0      14.0       8.0       515     3.0   
3                  90.0     77.0      34.0      34.0       515     3.0   
4                 115.0     87.0      11.0      25.0       688     4.0   
...                 ...      ...       ...       ...       ...     ...   
1936753           152.0    120.0      25.0      30.0       906     9.0   
1936754           109.0     78.0      75.0      57.0       481    15.0   
1936755           143.0    122.0      99.0      80.0       689     8.0   
1936756           117.0     89.0       9.0      11.0       533    13.0   
1936757           135.0    104.0      -5.0       7.0       874     8.0   

         TaxiOut  CarrierDelay  WeatherDelay   NASDelay  SecurityDelay  \
0            8.0     19.179399      3.703571  15.021635       0.090137   
1           10.0     19.179399      3.703571  15.021635       0.090137   
2           17.0     19.179399      3.703571  15.021635       0.090137   
3           10.0      2.000000      0.000000   0.000000       0.000000   
4           10.0     19.179399      3.703571  15.021635       0.090137   
...          ...           ...           ...        ...            ...   
1936753     18.0      3.000000      0.000000   0.000000       0.000000   
1936754     34.0      0.000000     57.000000  18.000000       0.000000   
1936755     32.0      1.000000      0.000000  19.000000       0.000000   
1936756     13.0     19.179399      3.703571  15.021635       0.090137   
1936757     11.0     19.179399      3.703571  15.021635       0.090137   

         LateAircraftDelay  
0                25.296466  
1                25.296466  
2                25.296466  
3                32.000000  
4                25.296466  
...                    ...  
1936753          22.000000  
1936754           0.000000  
1936755          79.000000  
1936756          25.296466  
1936757          25.296466  

[1936758 rows x 17 columns]

+ *Del Data set original, en faig un random sample de 2000 observacions, per millor agilitat de l'ordinador*

In [23]:
retards_df = retards_df.sample(n = 2000,  random_state = 4).reset_index(drop=True)
retards_df

Month  DayofMonth  DayOfWeek  FlightNum  ActualElapsedTime  \
0         4          24          4       1456               84.0   
1         5          26          1        354              153.0   
2         1          28          1       3802              125.0   
3         3           8          6       1449              158.0   
4        11          11          2        442              159.0   
...     ...         ...        ...        ...                ...   
1995      6          26          4        980              149.0   
1996      9          13          6       1144              198.0   
1997      4          22          2        372              134.0   
1998      2          12          2        341              156.0   
1999     11           5          3       2324               57.0   

      CRSElapsedTime  AirTime  ArrDelay  DepDelay  Distance  TaxiIn  TaxiOut  \
0               70.0     51.0      28.0      14.0       236     7.0     26.0   
1              177.0    130.0      -7.0      17.0      1024     5.0     18.0   
2               95.0     81.0      73.0      43.0       532     8.0     36.0   
3              168.0    131.0     437.0     447.0      1024    11.0     16.0   
4               86.0     57.0     130.0      57.0       429     7.0     95.0   
...              ...      ...       ...       ...       ...     ...      ...   
1995           155.0    136.0      80.0      86.0       935     5.0      8.0   
1996           205.0    178.0       9.0      16.0      1464     5.0     15.0   
1997           136.0    111.0      72.0      74.0       761    10.0     13.0   
1998           170.0    125.0     274.0     288.0       733    14.0     17.0   
1999            55.0     45.0      28.0      26.0       197     6.0      6.0   

      CarrierDelay  WeatherDelay    NASDelay  SecurityDelay  LateAircraftDelay  
0         0.000000      0.000000   14.000000       0.000000          14.000000  
1        19.179399      3.703571   15.021635       0.090137          25.296466  
2        73.000000      0.000000    0.000000       0.000000           0.000000  
3         0.000000      0.000000  437.000000       0.000000           0.000000  
4         0.000000      7.000000   73.000000       0.000000          50.000000  
...            ...           ...         ...            ...                ...  
1995      0.000000      0.000000    0.000000       0.000000          80.000000  
1996     19.179399      3.703571   15.021635       0.090137          25.296466  
1997      0.000000      0.000000    0.000000       0.000000          72.000000  
1998      0.000000      0.000000   63.000000       0.000000         211.000000  
1999      0.000000      0.000000    2.000000       0.000000          26.000000  

[2000 rows x 17 columns]

+ *Divideixo el Data Set a (variables independents) i b (variable dependent), només per tenir variable target al final del 
   DataFrame*

In [24]:
a = retards_df.drop("ArrDelay", 1)
b = retards_df["ArrDelay"]

+ *I faig un "pd.concat()" per deixar el Data set amb la variable dependent "ArrDelay" al final a la dreta*

In [25]:
retards2_df = pd.concat([a, b], axis = 1)

In [26]:
retards2_df

Month  DayofMonth  DayOfWeek  FlightNum  ActualElapsedTime  \
0         4          24          4       1456               84.0   
1         5          26          1        354              153.0   
2         1          28          1       3802              125.0   
3         3           8          6       1449              158.0   
4        11          11          2        442              159.0   
...     ...         ...        ...        ...                ...   
1995      6          26          4        980              149.0   
1996      9          13          6       1144              198.0   
1997      4          22          2        372              134.0   
1998      2          12          2        341              156.0   
1999     11           5          3       2324               57.0   

      CRSElapsedTime  AirTime  DepDelay  Distance  TaxiIn  TaxiOut  \
0               70.0     51.0      14.0       236     7.0     26.0   
1              177.0    130.0      17.0      1024     5.0     18.0   
2               95.0     81.0      43.0       532     8.0     36.0   
3              168.0    131.0     447.0      1024    11.0     16.0   
4               86.0     57.0      57.0       429     7.0     95.0   
...              ...      ...       ...       ...     ...      ...   
1995           155.0    136.0      86.0       935     5.0      8.0   
1996           205.0    178.0      16.0      1464     5.0     15.0   
1997           136.0    111.0      74.0       761    10.0     13.0   
1998           170.0    125.0     288.0       733    14.0     17.0   
1999            55.0     45.0      26.0       197     6.0      6.0   

      CarrierDelay  WeatherDelay    NASDelay  SecurityDelay  \
0         0.000000      0.000000   14.000000       0.000000   
1        19.179399      3.703571   15.021635       0.090137   
2        73.000000      0.000000    0.000000       0.000000   
3         0.000000      0.000000  437.000000       0.000000   
4         0.000000      7.000000   73.000000       0.000000   
...            ...           ...         ...            ...   
1995      0.000000      0.000000    0.000000       0.000000   
1996     19.179399      3.703571   15.021635       0.090137   
1997      0.000000      0.000000    0.000000       0.000000   
1998      0.000000      0.000000   63.000000       0.000000   
1999      0.000000      0.000000    2.000000       0.000000   

      LateAircraftDelay  ArrDelay  
0             14.000000      28.0  
1             25.296466      -7.0  
2              0.000000      73.0  
3              0.000000     437.0  
4             50.000000     130.0  
...                 ...       ...  
1995          80.000000      80.0  
1996          25.296466       9.0  
1997          72.000000      72.0  
1998         211.000000     274.0  
1999          26.000000      28.0  

[2000 rows x 17 columns]

In [27]:
retards2_df.describe()

Month   DayofMonth    DayOfWeek    FlightNum  ActualElapsedTime  \
count  2000.000000  2000.000000  2000.000000  2000.000000        2000.000000   
mean      6.039000    15.787500     3.973500  2145.272500         144.941844   
std       3.510932     8.799378     1.994439  1947.304349         136.384458   
min       1.000000     1.000000     1.000000     1.000000          30.000000   
25%       3.000000     8.000000     2.000000   608.500000          81.000000   
50%       6.000000    16.000000     4.000000  1472.000000         116.000000   
75%       9.000000    23.000000     6.000000  3330.000000         170.250000   
max      12.000000    31.000000     7.000000  7767.000000        1610.140629   

       CRSElapsedTime      AirTime    DepDelay     Distance       TaxiIn  \
count     2000.000000  2000.000000  2000.00000  2000.000000  2000.000000   
mean       136.589500   110.991663    45.00950   783.710000     6.966471   
std         74.096593    71.574275    57.62573   595.294679     6.180669   
min         34.000000    15.000000     6.00000    67.000000     1.000000   
25%         83.000000    59.000000    12.00000   349.750000     4.000000   
50%        116.000000    91.000000    25.00000   611.500000     5.000000   
75%        169.000000   141.000000    55.00000  1024.000000     8.000000   
max        635.000000   599.000000  1153.00000  4962.000000   121.000000   

           TaxiOut  CarrierDelay  WeatherDelay     NASDelay  SecurityDelay  \
count  2000.000000   2000.000000   2000.000000  2000.000000    2000.000000   
mean     18.148616     18.334751      3.687639    16.618508       0.149778   
std      13.725838     34.898253     15.250171    31.772982       3.022160   
min       3.000000      0.000000      0.000000     0.000000       0.000000   
25%      10.000000      0.000000      0.000000     0.000000       0.000000   
50%      14.000000     19.179399      0.000000    15.021635       0.000000   
75%      21.000000     19.179399      3.703571    15.021635       0.090137   
max     173.000000    865.000000    316.000000   437.000000     126.000000   

       LateAircraftDelay     ArrDelay  
count        2000.000000  2000.000000  
mean           26.133874    44.550199  
std            36.896660    60.926511  
min             0.000000   -44.000000  
25%             0.000000     9.750000  
50%            25.296466    26.000000  
75%            25.296466    58.000000  
max           470.000000  1129.000000

+ *Comprovo que el shape de "X" i de "y", ha quedat correctament*

+ *Divideixo el Data Set en "X" (variables independents) i "y" (variable dependent)*

+ *I Comprovo que el shape de "X" i de "y", ha quedat correctament*

In [28]:
X = retards2_df.drop("ArrDelay", 1)
y = retards2_df["ArrDelay"]

In [29]:
X.shape

(2000, 16)

In [30]:
type(X)

pandas.core.frame.DataFrame

In [31]:
y.shape

(2000,)

In [32]:
type(y)

pandas.core.series.Series

### a.1) OPCIÓ AMB MULTILINEAR REGRESSION - AMB TOTES LES DADES

In [34]:
from sklearn.linear_model import LinearRegression

+ *Definim el model de regressió*

In [35]:
model = LinearRegression().fit(X, y)

+ *L'entrenem amb ".fit" sobre les variables independents*

+ *Obtenim R2 - coeficient de determinació*

In [36]:
r_sq = model.score(X, y)
print('coefficient of determination:', r_sq)

coefficient of determination: 0.9939342128701782


+ *Fem una predicció amb ".predict(X)" - on X = variables independents -*

In [37]:
y_predict = model.predict(X)
print(y_predict)

[ 27.35587026  -4.31769317  72.49846559 ...  71.52473581 274.18715038
  26.10304119]


+ *Els valors reals del data set per "ArrDelay" són [ 28.0,  -7.0,  73.0, ...  72.0, 274.0,  28.0]*

  *D'acord amb R2 = 0.9940, podem veure con s'apropa el model quan fa la seva predicció, als valors reals que tenim*

+ *Importo "mean_squared_error" per utilitzar-lo*

In [38]:
from sklearn.metrics import mean_squared_error

In [39]:
print('MSE = {}'.format (mean_squared_error(y, y_predict)))

MSE = 22.50518467751974


+ *Hem obtingut un MSE de 22.50*

+ *I per últim obtenim "intercept" i els "coefficients" de vcada variable independen de l'equació de regressió.*

In [40]:
print('intercept:', model.intercept_)
print('coefficients:', model.coef_)

intercept: -1.4690977931443427
coefficients: [-4.86417823e-02  1.68501795e-02 -1.64339344e-02  6.55058004e-05
  3.36204698e-04 -9.22905582e-01  8.90147380e-01  9.80383094e-01
  3.88340440e-03  9.84537603e-01  9.85981538e-01  1.71201039e-02
  1.71252738e-02  2.40504687e-02  2.24576580e-02  1.83776300e-02]


### a.2) OPCIÓ AMB STATS MODELS - AMB TOTES LES DADES - (alternativa al model anterior.)

In [41]:
import statsmodels.api as sm

In [42]:
X = sm.add_constant(X)

In [43]:
model = sm.OLS(y, X)

In [44]:
results = model.fit()

In [45]:
 print(results.summary())

                            OLS Regression Results                            
Dep. Variable:               ArrDelay   R-squared:                       0.994
Model:                            OLS   Adj. R-squared:                  0.994
Method:                 Least Squares   F-statistic:                 2.031e+04
Date:                Wed, 30 Jun 2021   Prob (F-statistic):               0.00
Time:                        21:28:04   Log-Likelihood:                -5951.6
No. Observations:                2000   AIC:                         1.194e+04
Df Residuals:                    1983   BIC:                         1.203e+04
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -1.4691      0.55

+ *Veiem en el quadre del summary que R2 = 0.994*

  *Però el calculem amb el mètode que inclou el mòdul statsmodels.api*

In [46]:
print('R2 = {}'.format (results.rsquared))

R2 = 0.9939342128701782


+ *I veiem que és exactament igual a l'obtingut amb Multilinear Regression*

+ *Fem la predicció amb el model definit*

In [47]:
predictions = results.predict()

In [48]:
predictions

array([ 27.35587026,  -4.31769317,  72.49846559, ...,  71.52473581,
       274.18715038,  26.10304119])

+ *Els valors reals del data set per "ArrDelay" són [ 28.  -7.  73. ...  72. 274.  28.]*

  *D'acord amb R2 = 0.9940, podem veure con s'apropa el model quan fa la seva predicció, als valors reals que tenim*

In [49]:
sm.tools.eval_measures.mse(y, predictions, axis=0)

22.505184677519747

+ *Veiem que el MSE és el mateix que hem obtingut amb Multilinear Regression*

### b) OPCIÓ AMB POLYNOMIAL FEATURES - AMB TOTES LES DADES

In [50]:
from sklearn.preprocessing import PolynomialFeatures

+ *Fixem el grau = 2, de la funció polinòmica de regressió (encara que és el valor del paràmetre degree per Default)*

In [51]:
poly_reg = PolynomialFeatures(degree=2)

In [52]:
poly_reg.fit(X)

PolynomialFeatures()

+ *I apliquem ".transform()" per transformar les dades*

In [53]:
X_poly = poly_reg.transform(X)

In [54]:
type(X_poly)

numpy.ndarray

In [56]:
X_poly

array([[1.00000000e+00, 1.00000000e+00, 4.00000000e+00, ...,
        0.00000000e+00, 0.00000000e+00, 1.96000000e+02],
       [1.00000000e+00, 1.00000000e+00, 5.00000000e+00, ...,
        8.12470393e-03, 2.28015110e+00, 6.39911201e+02],
       [1.00000000e+00, 1.00000000e+00, 1.00000000e+00, ...,
        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],
       ...,
       [1.00000000e+00, 1.00000000e+00, 4.00000000e+00, ...,
        0.00000000e+00, 0.00000000e+00, 5.18400000e+03],
       [1.00000000e+00, 1.00000000e+00, 2.00000000e+00, ...,
        0.00000000e+00, 0.00000000e+00, 4.45210000e+04],
       [1.00000000e+00, 1.00000000e+00, 1.10000000e+01, ...,
        0.00000000e+00, 0.00000000e+00, 6.76000000e+02]])

In [64]:
len(X_poly)

2000

In [57]:
type(y)

pandas.core.series.Series

In [58]:
model = LinearRegression()

In [59]:
model.fit(X_poly, y)

LinearRegression()

In [60]:
r_sq = model.score(X_poly, y)
print('coefficient of determination:', r_sq)

coefficient of determination: 0.9999999845768733


In [61]:
y_predict = model.predict(X_poly)
print(y_predict)

[ 27.99968696  -6.98310542  73.00294066 ...  72.00023298 273.99945955
  28.00283237]


+ *Els valors reals del data set per "ArrDelay" són [ 28.  -7.  73. ...  72. 274.  28.]*

  *D'acord amb R2 = 9,999999 podem veure com el model encerta quasi perfectament quan fa la seva predicció, els valors reals que tenim*

In [62]:
print('MSE = {}'.format (mean_squared_error(y, y_predict)))

MSE = 5.722263363604194e-05


+ *+ *I en consequència, obtenim un MSE quasi igual a zero = 5.722263363604194e-05*

In [63]:
print('intercept:', model.intercept_)
print('coefficients:', model.coef_)

intercept: 0.0024156391464487115
coefficients: [ 1.73249908e-09  6.11715355e-09  9.85326389e-05 -6.19279609e-05
 -3.10630896e-04 -3.44609491e-07  2.83006544e-01 -3.77372099e-01
  9.44927199e-02  3.77290671e-01 -1.55994175e-05  9.42190778e-02
  9.42957061e-02  1.22724573e-01  1.22703726e-01  1.22687872e-01
  5.40100446e-01  1.22712697e-01  1.53734794e-11  9.85315672e-05
 -6.19292967e-05 -3.10631267e-04 -3.45127776e-07  2.83006545e-01
 -3.77372099e-01  9.44927198e-02  3.77290671e-01 -1.55994093e-05
  9.42190777e-02  9.42957061e-02  1.22724572e-01  1.22703726e-01
  1.22687872e-01  5.40100446e-01  1.22712697e-01 -2.74111129e-06
  5.65896813e-06  4.80195014e-05  1.58492795e-08  1.15792186e-04
  1.64705320e-04 -2.84327739e-04 -1.76513595e-04  4.21720045e-07
 -2.86899723e-04 -2.95340226e-04  1.76516066e-04  1.78455789e-04
  1.76032731e-04 -1.19378614e-01  1.77017592e-04  2.83120786e-06
  2.72333086e-06  5.65982898e-09  4.87309786e-05  6.01287259e-05
 -1.11474552e-04 -6.37882287e-05  3.1020962

### c) OPCIÓ AMB DECISION TREE REGRESSION - AMB TOTES LES DADES

In [65]:
from sklearn.tree import DecisionTreeRegressor

In [66]:
regressor = DecisionTreeRegressor(random_state=0)

In [67]:
regressor.fit(X,y)

DecisionTreeRegressor(random_state=0)

+ *Obtenim R2*

In [68]:
r_sq = regressor.score(X, y)
print('coefficient of determination:', r_sq)

coefficient of determination: 1.0


+ *Fem la predicció sobre X*

In [69]:
y_predict =regressor.predict(X)
print(y_predict)

[ 28.  -7.  73. ...  72. 274.  28.]


+ *Els valors reals del data set per "ArrDelay" són [ 28.  -7.  73. ...  72. 274.  28.]*

  *D'acord amb R2 = 1.0, podem veure com el model encerta perfectament quan fa la seva predicció, els valors reals que tenim*

+ *I obtenim el MSE*

In [70]:
print('MSE = {}'.format (mean_squared_error(y, y_predict)))

MSE = 0.0


+ *I d'acord amb el R2 obtingut = 1.0; MSE = 0.0

### RESUM COMPARACIÓ MODELS REGRESSIÓ UTILITZATS



Amb tantes features (17) i amb 2000 observacions, els models de Multilinear Regression i el de "statsmodels" han donat mateixos resultats de R2 i MSE:
+ coefficient of determination: 0.9939
+ MSE = 22.50

Els models anteriors han donat pitjors resultats que  els models de Linear Regression amb Polynomial Features i Decision Tree Regression que han estat respectivament:
+ coefficient of determination: 0.9999 i  MSE =  5.722263363604194e-05
+ coefficient of determination: 1.0 i  MSE = 0.0.

El que jo entenc, és que estem davant d'un cas clar de "overfitting".La quantitat de features utilitzades per l'entrenament (17) i la quantitat d'observacions contemplades pels models de regressió (2000), fan que  obtinguem valors de R2 molt alts, amb multilinear regression (i Statsmodel) però probablement no ho farien tant bé amb new Data amb les que fer predicció.

El mateix argument serveix per argumentar que el models Polynomial Features i Decission Tree Regression, aconsegueixen una R2 = 1 (o quasi).
I en consequència un MSE igual o quasi zero i igual a 0.0 respectivament.

###  *Exercici 3*


Entrena’ls utilitzant els diferents paràmetres que admeten

In [71]:
retards2_df

Month  DayofMonth  DayOfWeek  FlightNum  ActualElapsedTime  \
0         4          24          4       1456               84.0   
1         5          26          1        354              153.0   
2         1          28          1       3802              125.0   
3         3           8          6       1449              158.0   
4        11          11          2        442              159.0   
...     ...         ...        ...        ...                ...   
1995      6          26          4        980              149.0   
1996      9          13          6       1144              198.0   
1997      4          22          2        372              134.0   
1998      2          12          2        341              156.0   
1999     11           5          3       2324               57.0   

      CRSElapsedTime  AirTime  DepDelay  Distance  TaxiIn  TaxiOut  \
0               70.0     51.0      14.0       236     7.0     26.0   
1              177.0    130.0      17.0      1024     5.0     18.0   
2               95.0     81.0      43.0       532     8.0     36.0   
3              168.0    131.0     447.0      1024    11.0     16.0   
4               86.0     57.0      57.0       429     7.0     95.0   
...              ...      ...       ...       ...     ...      ...   
1995           155.0    136.0      86.0       935     5.0      8.0   
1996           205.0    178.0      16.0      1464     5.0     15.0   
1997           136.0    111.0      74.0       761    10.0     13.0   
1998           170.0    125.0     288.0       733    14.0     17.0   
1999            55.0     45.0      26.0       197     6.0      6.0   

      CarrierDelay  WeatherDelay    NASDelay  SecurityDelay  \
0         0.000000      0.000000   14.000000       0.000000   
1        19.179399      3.703571   15.021635       0.090137   
2        73.000000      0.000000    0.000000       0.000000   
3         0.000000      0.000000  437.000000       0.000000   
4         0.000000      7.000000   73.000000       0.000000   
...            ...           ...         ...            ...   
1995      0.000000      0.000000    0.000000       0.000000   
1996     19.179399      3.703571   15.021635       0.090137   
1997      0.000000      0.000000    0.000000       0.000000   
1998      0.000000      0.000000   63.000000       0.000000   
1999      0.000000      0.000000    2.000000       0.000000   

      LateAircraftDelay  ArrDelay  
0             14.000000      28.0  
1             25.296466      -7.0  
2              0.000000      73.0  
3              0.000000     437.0  
4             50.000000     130.0  
...                 ...       ...  
1995          80.000000      80.0  
1996          25.296466       9.0  
1997          72.000000      72.0  
1998         211.000000     274.0  
1999          26.000000      28.0  

[2000 rows x 17 columns]

In [72]:
retards2_df.columns

Index(['Month', 'DayofMonth', 'DayOfWeek', 'FlightNum', 'ActualElapsedTime',
       'CRSElapsedTime', 'AirTime', 'DepDelay', 'Distance', 'TaxiIn',
       'TaxiOut', 'CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay',
       'LateAircraftDelay', 'ArrDelay'],
      dtype='object')

+ *Divideixo el Data set  en X = variables independents i y = variable dependent o target*

In [74]:
X = retards2_df.drop("ArrDelay", 1)
y = retards2_df["ArrDelay"]

+ *Comprovo que el shape de "X" i de "y", ha quedat correctament*

In [75]:
X.shape

(2000, 16)

X = np.array(X)

In [76]:
type(X)

pandas.core.frame.DataFrame

In [78]:
y.shape

(2000,)

In [79]:
type(y)

pandas.core.series.Series

### 1.A- MULTILINEAR REGRESSION amb fit_intercept=False

In [80]:
from sklearn.linear_model import LinearRegression

+ *Definim el model de regressió*

In [81]:
model = LinearRegression(fit_intercept = False)

+ *L'entrenem amb ".fit" sobre les variables independents*

In [82]:
model.fit(X, y)

LinearRegression(fit_intercept=False)

+ *Obtenim R2 - coeficient de determinació*

In [83]:
r_sq = model.score(X, y)
print('coefficient of determination:', r_sq)

coefficient of determination: 0.9939126461429176


+ *Fem una predicció amb ".predict(X)" - on X = variables independents -*

In [84]:
y_predict = model.predict(X)
print(y_predict)

[ 27.53387754  -4.33890984  72.7740393  ...  71.73426499 273.93348847
  26.50184589]


+ *Els valors reals del data set per "ArrDelay" són [ 28.  -7.  73. ...  72. 274.  28.]*

  *D'acord amb R2 = 0.9449, podem veure con s'apropa el model quan fa la seva predicció, als valors reals que tenim*

+ *Importo "mean_squared_error" per utilitzar-lo*

In [85]:
from sklearn.metrics import mean_squared_error

In [86]:
print('MSE = {}'.format (mean_squared_error(y, y_predict)))

MSE = 22.58520119796513


+ *Hem obtingut un MSE de 22.58520119796513*

In [87]:
print('intercept:', model.intercept_)
print('coefficients:', model.coef_)

intercept: 0.0
coefficients: [-7.25254927e-02  5.95973375e-03 -7.55610068e-02  1.94476155e-05
  3.80421666e-04 -9.36196374e-01  8.91829496e-01  9.82408633e-01
  5.14001231e-03  9.80213280e-01  9.85546116e-01  1.41753824e-02
  1.46708431e-02  2.20439376e-02  2.23465126e-02  1.51635578e-02]


+ *__Observem com ara "intercept = 0" tal com fixa el paràmetre opcional utilitzat, (fit_intercept=False)__*

### 1.B- MULTILINEAR REGRESSION amb normalize = True

In [88]:
from sklearn.linear_model import LinearRegression

+ *Definim el model de regressió*

In [89]:
model = LinearRegression(normalize = True)

+ *L'entrenem amb ".fit" sobre les variables independents*

In [90]:
model.fit(X, y)

LinearRegression(normalize=True)

+ *Obtenim R2 - coeficient de determinació*

In [91]:
r_sq = model.score(X, y)
print('coefficient of determination:', r_sq)

coefficient of determination: 0.9939342128701782


+ *Fem una predicció amb ".predict(X)" - on X = variables independents -*

In [92]:
y_predict = model.predict(X)
print(y_predict)

[ 27.35587026  -4.31769317  72.49846559 ...  71.52473581 274.18715038
  26.10304119]


+ *Els valors reals del data set per "ArrDelay" són [ 28.  -7.  73. ...  72. 274.  28.]*

  *D'acord amb R2 = 0.9693, podem veure con s'apropa el model quan fa la seva predicció, als valors reals que tenim*

+ *Importo "mean_squared_error" per utilitzar-lo*

In [93]:
from sklearn.metrics import mean_squared_error

In [94]:
print('MSE = {}'.format (mean_squared_error(y, y_predict)))

MSE = 22.505184677519736


+ *Hem obtingut un MSE de 113.82631621704608*

In [95]:
print('intercept:', model.intercept_)
print('coefficients:', model.coef_)

intercept: -1.4690977931443285
coefficients: [-4.86417823e-02  1.68501795e-02 -1.64339344e-02  6.55058004e-05
  3.36204698e-04 -9.22905582e-01  8.90147380e-01  9.80383094e-01
  3.88340440e-03  9.84537603e-01  9.85981538e-01  1.71201039e-02
  1.71252738e-02  2.40504687e-02  2.24576580e-02  1.83776300e-02]


## SUMMARY OPCIÓ 1 : MULTILINEAR REGRESSION

+ *Podem veure que el coefficient of determination  i MSE han sigut idèntics per el model amb el paràmetres per default i
   amb "normalize = True" (R2 = 0.9939342 i MSE = 22.50)
   
   *En canvi quan hem definit "Intercept = False" com a a paràmetre, el resultats han canviat una mica (poc) i hem obtingut
     R2 = 0.993912 i MSE = 22.58.*

### 2.A.- POLYNOMIAL FEATURES amb degree=3

+ *El paràmetre "degree" és igual a 2 per Default (de fet és el mínim grau que ha de tenir l'equació de regressió per poder 
   ser considerada polinònica)*
   
   *El que faré ara, és comprovar com pot millorar el model si fixo el degree = 3*

In [96]:
from sklearn.preprocessing import PolynomialFeatures

+ *Fixem el grau = 3, de la funció polinòmica de regressió*

In [97]:
poly_reg = PolynomialFeatures(degree=3)

In [98]:
poly_reg.fit(X)

PolynomialFeatures(degree=3)

+ *I apliquem ".transform()" per transformar les dades*

In [99]:
X_poly = poly_reg.transform(X)

In [100]:
type(X_poly)

numpy.ndarray

In [101]:
X_poly

array([[1.00000000e+00, 4.00000000e+00, 2.40000000e+01, ...,
        0.00000000e+00, 0.00000000e+00, 2.74400000e+03],
       [1.00000000e+00, 5.00000000e+00, 2.60000000e+01, ...,
        2.05526298e-01, 5.76797653e+01, 1.61874921e+04],
       [1.00000000e+00, 1.00000000e+00, 2.80000000e+01, ...,
        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],
       ...,
       [1.00000000e+00, 4.00000000e+00, 2.20000000e+01, ...,
        0.00000000e+00, 0.00000000e+00, 3.73248000e+05],
       [1.00000000e+00, 2.00000000e+00, 1.20000000e+01, ...,
        0.00000000e+00, 0.00000000e+00, 9.39393100e+06],
       [1.00000000e+00, 1.10000000e+01, 5.00000000e+00, ...,
        0.00000000e+00, 0.00000000e+00, 1.75760000e+04]])

In [104]:
model = LinearRegression()

In [105]:
model.fit(X_poly, y)

LinearRegression()

In [106]:
r_sq = model.score(X_poly, y)
print('coefficient of determination:', r_sq)

coefficient of determination: 0.9999999999999986


In [107]:
y_predict = model.predict(X_poly)
print(y_predict)

[ 28.00000074  -7.00000258  72.99999871 ...  72.00000001 273.99999917
  28.00000107]


+ *Els valors reals del data set per "ArrDelay" són [ 28.  -7.  73. ...  72. 274.  28.]*

  *D'acord amb R2 = 0.9999, podem veure com s'apropa el model quan fa la seva predicció, als valors reals que tenim*

In [108]:
print('MSE = {}'.format (mean_squared_error(y, y_predict)))

MSE = 5.556444774314421e-12


+  *I obtenim un MSE = 5.556444774314421e-12 quasi igual a zero*

In [109]:
print('intercept:', model.intercept_)
print('coefficients:', model.coef_)

intercept: -5.15042970050672e-07
coefficients: [ 1.40215064e-06 -1.52633221e-07 -1.03758239e-07  2.95236482e-09
  8.42619607e-10  4.44444887e-01 -5.92593165e-01  1.48148261e-01
  5.92593099e-01  1.49267614e-09  1.48148290e-01  1.48148217e-01
  4.07406940e-01  4.07406940e-01  4.07406884e-01  8.36932927e-08
  4.07406903e-01 -5.35011452e-09 -1.25666396e-09 -6.94626710e-09
  1.91929002e-11  4.24731810e-08 -5.34386894e-08  1.50469998e-08
  5.58976514e-08 -3.65326541e-10  1.42841287e-08  1.57968961e-08
 -5.47775160e-08 -7.97927113e-08 -5.74120694e-08  6.05682317e-07
 -5.53406287e-08  1.13349191e-09  1.03423367e-08  8.43297508e-12
  1.22086719e-07 -1.60437408e-07  3.93762924e-08  1.61602567e-07
 -7.40233415e-11  4.02455261e-08  4.10687168e-08 -1.62494103e-07
 -1.59105965e-07 -1.62159340e-07  1.52993728e-06 -1.61900340e-07
 -1.95451994e-08  2.83405486e-12  5.19438758e-08 -6.35703152e-08
  1.20535317e-08  6.85452943e-08  1.09245855e-10  1.99357606e-08
  2.05442129e-08 -6.78081684e-08 -9.6642252

In [110]:
len(model.coef_)

969

### 2.B- POLYNOMIAL FEATURES amb interaction_only = True

+ *El paràmetre "interaction_only" és igual a False per Default.Set to True vol dir que volem que inclogui només les 
  interaction features que es generen en la confecció de l'equació de regressió (i NO totes les features).
   

In [111]:
from sklearn.preprocessing import PolynomialFeatures

+ *Fixem el grau = 2, de la funció polinòmica de regressió per fer vàlida la comparació amb el cas que, amb degree = 2,
   NO haviem fixat cap paràmetre opcional*

In [112]:
poly_reg = PolynomialFeatures(degree=2, interaction_only = True)

In [113]:
poly_reg.fit(X)

PolynomialFeatures(interaction_only=True)

+ *I apliquem ".transform()" per transformar les dades*

In [114]:
X_poly = poly_reg.transform(X)

In [115]:
type(X_poly)

numpy.ndarray

In [116]:
model = LinearRegression()

In [117]:
model.fit(X_poly, y)

LinearRegression()

In [118]:
r_sq = model.score(X_poly, y)
print('coefficient of determination:', r_sq)

coefficient of determination: 0.9999996360022797


In [119]:
y_predict = model.predict(X_poly)
print(y_predict)

[ 28.00757713  -6.89976919  72.98336853 ...  72.0055469  274.05722955
  27.97734651]


+ *Els valors reals del data set per "ArrDelay" són [ 28.  -7.  73. ...  72. 274.  28.]*

  *D'acord amb R2 = 0.9764763223554763, podem veure com s'apropa el model  quan fa la seva predicció, als valors reals que 
   tenim*

In [120]:
print('MSE = {}'.format (mean_squared_error(y, y_predict)))

MSE = 0.0013504984171754451


+  *I obtenim un MSE =  0.0013504984171754451*

In [121]:
print('intercept:', model.intercept_)
print('coefficients:', model.coef_)

intercept: -0.04347595849966979
coefficients: [ 1.45887296e-09 -8.29104936e-04  1.80177042e-03  1.23934141e-03
 -3.30447475e-06  7.49760960e-01 -9.99390670e-01  2.48648994e-01
  1.00356525e+00  1.17582026e-04  2.50483458e-01  2.53240694e-01
 -3.76049692e-03 -3.42116693e-03 -2.61844769e-03  9.71694112e+00
 -3.67959659e-03 -7.82774829e-05  9.50122094e-05  1.62221663e-07
  1.37264890e-04 -8.10127894e-05 -5.68149728e-06  9.06923314e-05
 -5.82708144e-06  4.14560451e-05 -6.27568723e-05 -9.17464233e-05
 -8.85156121e-05 -1.03258005e-04  6.22768132e-02 -9.33917455e-05
  1.38155680e-05 -3.85185277e-08  4.36610284e-05  4.69631782e-04
 -5.18473779e-04 -5.37432064e-04  7.18391861e-07 -5.14151530e-04
 -5.72541487e-04  5.38878453e-04  5.41346597e-04  5.27187621e-04
 -3.55112200e-01  5.38584019e-04 -8.33874978e-08 -1.73362072e-04
 -9.35850466e-04  1.16447706e-03  8.94032216e-04 -6.24390396e-06
  1.02574796e-03  9.83565542e-04 -8.84196302e-04 -8.98296025e-04
 -8.88143270e-04  5.60873895e-01 -9.00463126

In [122]:
len(model.coef_)

137

__*Veiem que quan hem fet el print de "coefficients" obtenim 137 coeficients en lloc dels 969 de quan no hem fixat
   el paràmetre interaction_only = True__*

## SUMMARY OPCIÓ POLYNOMIAL FEATURES (MULTILINEAR REGRESSION)
 
+ *Els resultats per P.Features amb paràmetres per default, amb "degree = 3* i amb "interaction_only = True"; han estat  
   respectivament :
   
   + R2 = 0.9999999845768733 i MSE = 5.556444774314421e-12.
   + R2 = 0.9999999999999986 i MSE = 5.556444774314421e-12.
   + R2 = 0.9999996360022797 i MSE = 0.0013504984171754451
   
  

### 3.- DECISION TREE REGRESSION amb criterion = "mae" ("mse" per Default) i splittter = "random"

+ *El paràmetre "criterion" permet escollir el mètode de selecció d'atribut per iniciar l'arbre de decissió i splitter
   defineix l'estratègia per escollir l'split a cada node *.
   
   *El que faré ara, és comprovar si pot millorar el model*

In [123]:
from sklearn.tree import DecisionTreeRegressor

In [124]:
regressor.fit(X,y)

DecisionTreeRegressor(random_state=0)

In [125]:
r_sq = regressor.score(X, y)
print('coefficient of determination:', r_sq)


coefficient of determination: 1.0


In [126]:
y_predict =regressor.predict(X)
print(y_predict)

[ 28.  -7.  73. ...  72. 274.  28.]


+ *Els valors reals del data set per "ArrDelay" són [ 28.  -7.  73. ...  72. 274.  28.]*

  *D'acord amb R2 = 1.0, podem veure com el model encerta perfectament quan fa la seva predicció, els valors reals que tenim*

In [127]:
print('MSE = {}'.format (mean_squared_error(y, y_predict)))

MSE = 0.0


## SUMMARY OPCIÓ 3
 
+ *Els resultats per Decission Tree Regressor amb paràmetres per default; i amb criterion = "mae" i "splittter=random"; han
     estat respectivament :
   
   + R2 = 1.0 i MSE = 0.0.
   + R2 = 1.0 i MSE = 0.0.
   
   *És a dir, no han canviat gens:

### *Exercici 4*

Compara el seu rendiment utilitzant l’aproximació traint/test o utilitzant totes les dades (validació interna)

In [128]:
retards2_df

Month  DayofMonth  DayOfWeek  FlightNum  ActualElapsedTime  \
0         4          24          4       1456               84.0   
1         5          26          1        354              153.0   
2         1          28          1       3802              125.0   
3         3           8          6       1449              158.0   
4        11          11          2        442              159.0   
...     ...         ...        ...        ...                ...   
1995      6          26          4        980              149.0   
1996      9          13          6       1144              198.0   
1997      4          22          2        372              134.0   
1998      2          12          2        341              156.0   
1999     11           5          3       2324               57.0   

      CRSElapsedTime  AirTime  DepDelay  Distance  TaxiIn  TaxiOut  \
0               70.0     51.0      14.0       236     7.0     26.0   
1              177.0    130.0      17.0      1024     5.0     18.0   
2               95.0     81.0      43.0       532     8.0     36.0   
3              168.0    131.0     447.0      1024    11.0     16.0   
4               86.0     57.0      57.0       429     7.0     95.0   
...              ...      ...       ...       ...     ...      ...   
1995           155.0    136.0      86.0       935     5.0      8.0   
1996           205.0    178.0      16.0      1464     5.0     15.0   
1997           136.0    111.0      74.0       761    10.0     13.0   
1998           170.0    125.0     288.0       733    14.0     17.0   
1999            55.0     45.0      26.0       197     6.0      6.0   

      CarrierDelay  WeatherDelay    NASDelay  SecurityDelay  \
0         0.000000      0.000000   14.000000       0.000000   
1        19.179399      3.703571   15.021635       0.090137   
2        73.000000      0.000000    0.000000       0.000000   
3         0.000000      0.000000  437.000000       0.000000   
4         0.000000      7.000000   73.000000       0.000000   
...            ...           ...         ...            ...   
1995      0.000000      0.000000    0.000000       0.000000   
1996     19.179399      3.703571   15.021635       0.090137   
1997      0.000000      0.000000    0.000000       0.000000   
1998      0.000000      0.000000   63.000000       0.000000   
1999      0.000000      0.000000    2.000000       0.000000   

      LateAircraftDelay  ArrDelay  
0             14.000000      28.0  
1             25.296466      -7.0  
2              0.000000      73.0  
3              0.000000     437.0  
4             50.000000     130.0  
...                 ...       ...  
1995          80.000000      80.0  
1996          25.296466       9.0  
1997          72.000000      72.0  
1998         211.000000     274.0  
1999          26.000000      28.0  

[2000 rows x 17 columns]

In [129]:
X = retards2_df.drop("ArrDelay", 1)
y = retards2_df["ArrDelay"]

In [130]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state = 4)

+ *Comprovem shapes de cada sub dataset per veure que han quedat com necessitem*

In [131]:
X_train.shape, y_train.shape, X_test.shape, y_test.shape

((1600, 16), (1600,), (400, 16), (400,))

In [132]:
type(X_train)

pandas.core.frame.DataFrame

In [133]:
type(y_train)

pandas.core.series.Series

### TRAIN TEST AMB MULTILINEAR REGRESSION

In [134]:
from sklearn.linear_model import LinearRegression

+ *Definim el model de regressió*

In [135]:
model = LinearRegression()

+ *L'entrenem amb ".fit" sobre el train set*

In [136]:
model.fit(X_train, y_train)

LinearRegression()

+ *Obtenim R2 - coeficient de determinació amb X_train i y_train*

In [137]:
r_sq = model.score(X_train, y_train)
print('coefficient of determination:', r_sq)

coefficient of determination: 0.992196151386683


+ *Fem una predicció amb ".predict(X_test)" - on X_test = variables independents NO conegudes pel model -*

In [138]:
y_predict = model.predict(X_test)
y_predict

array([ 2.29133361e+01,  1.58633744e+02,  1.88282559e+01,  1.37619055e+01,
        1.00009497e+02,  6.48640719e+00,  2.59586525e+01,  3.10233856e+01,
        1.40495342e+02,  3.83368779e+01,  7.37624430e+01,  5.73425366e+01,
        1.12140457e+01,  9.48660294e+01,  4.17852640e+02,  1.09192457e+01,
        7.47375516e+01,  1.46368168e+01, -5.07524063e+00,  7.04021218e+00,
        1.99924131e+01, -6.45162493e-01,  5.13301995e+01,  7.41053197e+00,
        1.57477266e+01,  2.92025510e+01, -8.08155802e+00,  3.28254756e+01,
        1.13048302e+01,  6.68203810e+01, -1.28021846e+01,  1.20501518e+01,
        4.40939954e+01,  1.67340400e+01,  1.47935648e+01,  1.08103492e+01,
        7.88602580e+01,  1.74520850e+01,  5.65827044e+01,  8.15007026e+00,
       -1.45330498e+01,  6.04248757e+01,  6.56021018e+01,  1.25387944e+01,
        2.02157694e+01,  3.08386399e+01,  5.42223182e+00,  5.21540521e+01,
        2.56960169e+01,  2.89873857e+01, -4.88092267e+00,  1.41511271e+01,
        7.78054966e+00,  

In [139]:
y_test

1893     23.0
898     158.0
1578     20.0
559      16.0
88       99.0
        ...  
570      21.0
376      46.0
1249     85.0
492      24.0
922       6.0
Name: ArrDelay, Length: 400, dtype: float64

+ *Els valors reals de y_test són [ 23.0, 158.0, 20.0, ...  85.0, 24.0, 6.0]*
  *Els valors  de y_predict són   [ 22.91, 158.63, 18.82, ...  83.80, 21.69, 8.73]*

  *D'acord amb R2 = 0.9676, podem veure con s'apropa el model quan fa la seva predicció, als valors reals de y_test que tenim
   per contrastar*

+ *Importo "mean_squared_error" per utilitzar-lo*

In [140]:
from sklearn.metrics import mean_squared_error

In [141]:
print('MSE = {}'.format (mean_squared_error(y_test, y_predict)))

MSE = 3.742132527208146


+ *Hem obtingut un MSE de 3.742132527208146*

In [142]:
print('intercept:', model.intercept_)
print('coefficients:', model.coef_)

intercept: -1.8140329693728887
coefficients: [-4.95808472e-02  2.09264290e-02 -2.33511590e-02  7.80978639e-05
  7.18778960e-04 -9.03946993e-01  8.60189328e-01  9.75654471e-01
  5.13902703e-03  9.78245778e-01  9.83511733e-01  2.01757051e-02
  2.02614691e-02  3.13181386e-02  2.14156182e-02  2.29177215e-02]


### COMPARATIVA TRAIN/TEST vs VALIDACIÓ INTERNA (amb totes les dades) - MULTILINEAR REGRESSION__*
   
                                            R2                          MSE
                                               
   
  
  
  + t&test set  (test_size=0.2)...............................0.992196151386683......................3.742132527208146
   
  + Validació interna............................................0.9939342128701782......................22.50

### TRAIN TEST AMB POLYNOMIAL FEATURES

In [143]:
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures

+ *Fixem el grau = 2, de la funció polinòmica de regressió (és el valor per Default)*

In [144]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state = 4)

In [145]:
X_train

Month  DayofMonth  DayOfWeek  FlightNum  ActualElapsedTime  \
875      12          29          1       1567              239.0   
1735      5          23          5         24               68.0   
1729      4          10          4       3097               93.0   
362      12          23          2       2900               78.0   
890       4           4          5       3149               58.0   
...     ...         ...        ...        ...                ...   
1921     12          30          2        864              120.0   
709      11          24          1        773              148.0   
439      12          15          1       1015              166.0   
174      10          17          5        402              145.0   
1146      4           8          2       4459              149.0   

      CRSElapsedTime  AirTime  DepDelay  Distance  TaxiIn  TaxiOut  \
875            246.0    220.0      50.0      1515     6.0     13.0   
1735            70.0     53.0       6.0       324     5.0     10.0   
1729           100.0     57.0      36.0       319    25.0     11.0   
362             80.0     68.0      48.0       337     4.0      6.0   
890             45.0     30.0      33.0       155     2.0     26.0   
...              ...      ...       ...       ...     ...      ...   
1921           122.0     87.0      32.0       581    14.0     19.0   
709            142.0    122.0      78.0       732     6.0     20.0   
439            137.0    108.0      46.0       665    13.0     45.0   
174            145.0    124.0      99.0      1017     5.0     16.0   
1146           145.0    120.0       7.0       865    12.0     17.0   

      CarrierDelay  WeatherDelay   NASDelay  SecurityDelay  LateAircraftDelay  
875      43.000000      0.000000   0.000000       0.000000           0.000000  
1735     19.179399      3.703571  15.021635       0.090137          25.296466  
1729      0.000000      0.000000   0.000000       0.000000          29.000000  
362       0.000000      0.000000   5.000000       0.000000          41.000000  
890       0.000000      0.000000  13.000000       0.000000          33.000000  
...            ...           ...        ...            ...                ...  
1921      0.000000      0.000000   0.000000       0.000000          30.000000  
709      39.000000      0.000000   6.000000       0.000000          39.000000  
439       0.000000      0.000000  29.000000       0.000000          46.000000  
174      99.000000      0.000000   0.000000       0.000000           0.000000  
1146     19.179399      3.703571  15.021635       0.090137          25.296466  

[1600 rows x 16 columns]

In [146]:
type(X_train)

pandas.core.frame.DataFrame

*Generm una instancia de PolynomialFeatures*

In [147]:
transformer = PolynomialFeatures(degree=2)

+ *Abans de transformar l'input (X_train), l'hem d'entrenar amb ".fit()"*

In [148]:
transformer.fit(X_train)

PolynomialFeatures()

+ *Ara sí podem crear un nou / modificat input*

In [149]:
X_train_poly = transformer.transform(X_train)

+ *Creem el model i l'entrenem*

In [150]:
model = LinearRegression()

+ *Podem evaluar R2 sobre el training set amb X_train_poly (NO amb X_train)*

In [152]:
model.fit(X_train_poly, y_train)  

LinearRegression()

In [153]:
r_sq = model.score(X_train_poly, y_train)
print('coefficient of determination:', r_sq)

coefficient of determination: 0.999999993502675


+ *Transformem X_test per poder fer la predicció amb les variables també transformades*

In [154]:
X_test_poly = transformer.transform(X_test)

In [155]:
y_test_predicted = model.predict(X_test_poly)
print(y_test_predicted)

[ 2.30008309e+01  1.58000259e+02  1.99998154e+01  1.60016869e+01
  9.89994531e+01  5.99838949e+00  2.60002580e+01  3.10009241e+01
  1.39001478e+02  3.80007285e+01  7.50009463e+01  6.00001600e+01
  9.99703368e+00  9.79959865e+01  4.19997500e+02  1.09951784e+01
  7.29992343e+01  1.59996862e+01 -7.00325040e+00  5.00087346e+00
  2.30011886e+01 -3.00660075e+00  5.09993436e+01  6.00129481e+00
  1.79997764e+01  2.99999242e+01 -9.96776213e+00  3.29994061e+01
  1.09935358e+01  6.69981479e+01 -1.60180487e+01  1.19922186e+01
  4.60007084e+01  1.80000238e+01  1.30021412e+01  1.10016422e+01
  7.90002626e+01  1.79996089e+01  5.89995567e+01  7.99728654e+00
 -1.90088888e+01  6.19995817e+01  6.59992734e+01  1.19992115e+01
  2.09993222e+01  2.99978358e+01  3.99651997e+00  5.30006607e+01
  2.90010335e+01  2.89990533e+01 -5.99661941e+00  1.30087426e+01
  6.99904878e+00  1.89997450e+01  1.06002894e+02 -8.01333200e+00
  2.09992028e+01  4.00168501e+00  6.19976226e+01  2.29997167e+01
  1.00338248e+00  4.60004

In [156]:
print('MSE = {}'.format (mean_squared_error(y_test, y_test_predicted)))

MSE = 27.883424550923333


+ *Em dona un MSE = 27.883424550923333*


## COMPARATIVA TRAIN/TEST vs VALIDACIÓ INTERNA (amb totes les dades) - POLYNOMIAL FEATURES
   
                                            R2                          MSE
                                               
     
  + t&test set  (test_size=0.2)...............................0.999999993502675......................27.883424550923333
   
  + Validació interna............................................0.9999999845768733......................5.556444774314421e-12

### OPCIÓ AMB DECISION TREE REGRESSION 

In [170]:
from sklearn.tree import DecisionTreeRegressor

* *Ara canvio el random_state (= 42 en lloc de = 4) per prevenir que y_train ja vingui entrenat d'anteriors models*

In [177]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state = 42)

In [178]:
X_train

Month  DayofMonth  DayOfWeek  FlightNum  ActualElapsedTime  \
968      10           8          3       1682               96.0   
240       3          21          5         92              234.0   
819       8          18          1       6035               98.0   
692       3          27          4       4342               71.0   
420       2          28          4       2337              172.0   
...     ...         ...        ...        ...                ...   
1130      3          21          5        438              226.0   
1294      4          16          3       2896              178.0   
860       3          10          1       1403              149.0   
1459      1           4          5       1218              149.0   
1126     10           9          4       6604              147.0   

      CRSElapsedTime  AirTime  DepDelay  Distance  TaxiIn  TaxiOut  \
968             93.0     67.0      13.0       432     5.0     24.0   
240            250.0    218.0      44.0      1989     5.0     11.0   
819             95.0     58.0       9.0       409     4.0     36.0   
692             70.0     45.0       6.0       296     4.0     22.0   
420            205.0    154.0      39.0      1166     7.0     11.0   
...              ...      ...       ...       ...     ...      ...   
1130           185.0    167.0      16.0      1086     6.0     53.0   
1294           165.0    166.0      11.0      1052     4.0      8.0   
860            165.0    136.0      25.0      1052     3.0     10.0   
1459           170.0    134.0       6.0      1105     4.0     11.0   
1126           154.0    130.0       7.0      1082     4.0     13.0   

      CarrierDelay  WeatherDelay   NASDelay  SecurityDelay  LateAircraftDelay  
968       1.000000      0.000000   3.000000       0.000000          12.000000  
240      28.000000      0.000000   0.000000       0.000000           0.000000  
819      19.179399      3.703571  15.021635       0.090137          25.296466  
692      19.179399      3.703571  15.021635       0.090137          25.296466  
420      19.179399      3.703571  15.021635       0.090137          25.296466  
...            ...           ...        ...            ...                ...  
1130      0.000000      0.000000  57.000000       0.000000           0.000000  
1294     11.000000      0.000000  13.000000       0.000000           0.000000  
860      19.179399      3.703571  15.021635       0.090137          25.296466  
1459     19.179399      3.703571  15.021635       0.090137          25.296466  
1126     19.179399      3.703571  15.021635       0.090137          25.296466  

[1600 rows x 16 columns]

In [179]:
regressor = DecisionTreeRegressor(random_state=0)

In [180]:
regressor.fit(X_train,y_train)

DecisionTreeRegressor(random_state=0)

In [181]:
r_sq = model.score(X_train, y_train)
print('coefficient of determination:', r_sq)

coefficient of determination: 0.9825402183768975


In [182]:
y_predict = model.predict(X_test)
print(y_predict)

[ 18.         149.           4.           1.           2.
 -10.          62.         154.          55.          16.
 203.         125.          10.           8.           2.
  35.          35.         156.          10.          45.
  -5.           3.          23.           7.          17.
  17.          -5.          25.          12.          23.
  -3.          52.         164.          17.          60.
   7.          48.         242.          42.          12.
  19.          82.          14.          26.           8.
  -7.           2.          39.           9.          -8.
  19.          66.          12.          39.          58.
  10.          11.           5.          14.          -4.
  12.          93.          90.          11.         -37.
 242.          24.           4.           6.          77.
  42.19988477   1.          -9.           3.          95.
 -10.         141.           9.          37.          25.
   9.         108.          29.         166.          14.
  12.         

+ *Els valors reals de y_test són [18.0, 149.0, 4.0,,, ... 11.0, 1.0, 4.0]*
  *I els valors de y_predict són  [18.0, 149.0,4.0, ...  11.0, 3.0, 4.0]*


In [183]:
y_test

1860     18.0
353     149.0
1333      4.0
905     -14.0
1289      2.0
        ...  
965      36.0
1284     21.0
1739     11.0
261       3.0
535      30.0
Name: ArrDelay, Length: 400, dtype: float64

In [184]:
print('MSE = {}'.format (mean_squared_error(y_test, y_predict)))

MSE = 37.08


## COMPARATIVA TRAIN/TEST vs VALIDACIÓ INTERNA (amb totes les dades) - 
## DECISSION TREE REGRESSOR
   
                                            R2                          MSE
                                               
     
  + t&test set  (test_size=0.2)...............................0.9825402183768975......................37.08
   
  + Validació interna............................................1.0.....................................................0.0